File: lab1/python/oop/service/solver.py
from math import sqrt
from typing import Set

from entity.coefficients import Coefficients
from entity.exceptions import NoSolutions


class SolverService:
    @staticmethod
    def solve(coefficients: Coefficients) -> Set[float]:
        d = coefficients.B ** 2 - 4 * coefficients.A * coefficients.C
        if d < 0:
            raise NoSolutions()
        roots: Set[float] = set()

        possible_roots = [
            (-coefficients.B - sqrt(d)) / (2 * coefficients.A),
            (-coefficients.B + sqrt(d)) / (2 * coefficients.A),
        ]

        for root in possible_roots:
            if root >= 0:
                roots.add(sqrt(root))
                roots.add(-sqrt(root))

        return roots

File: lab1/golang/service/solver.go
package service

import (
	"lab1/domain"
	"math"
)

type SolverService struct{}

func NewSolverService() *SolverService {
	return &SolverService{}
}

func (s *SolverService) Solve(coefficients domain.Coefficients) ([]float64, error) {
	rootSet := map[float64]struct{}{}

	d := math.Pow(coefficients.B, 2) - 4*coefficients.A*coefficients.C
	if d < 0 {
		return []float64{}, domain.ErrNoSolutions
	}

	possibleRoots := []float64{
		(-coefficients.B - math.Sqrt(d)) / (2 * coefficients.A),
		(-coefficients.B + math.Sqrt(d)) / (2 * coefficients.A),
	}

	for _, root := range possibleRoots {
		if root < 0 {
			continue
		}

		rootSet[-math.Sqrt(root)] = struct{}{}
		rootSet[math.Sqrt(root)] = struct{}{}
	}

	roots := make([]float64, 0, len(rootSet))
	for root := range rootSet {
		roots = append(roots, root)
	}

	return roots, nil
}
